ADBDEV-2504 6.19.3 sync

deart2k opened the pull request:
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

macdice committed:
Free disk space for dropped relations on commit.

When committing a transaction that dropped a relation, we previously
truncated only the first segment file to free up disk space (the one
that won't be unlinked until the next checkpoint).

Truncate higher numbered segments too, even though we unlink them on
commit.  This frees the disk space immediately, even if other backends
have open file descriptors and might take a long time to get around to
handling shared invalidation events and closing them.  Also extend the
same behavior to the first segment, in recovery.

Back-patch to all supported releases.

Bug: #16663
Reported-by: Denis Patron <denis.patron@previnet.it>
Reviewed-by: Pavel Borisov <pashkin.elfe@gmail.com>
Reviewed-by: Neil Chen <carpenter.nail.cz@gmail.com>
Reviewed-by: David Zhang <david.zhang@highgo.ca>
Discussion: https://postgr.es/m/16663-fe97ccf9932fc800%40postgresql.org

See Greenplum PR #11988 for more details.

Backported from GPDB master:
https://github.com/greenplum-db/gpdb/commit/d9d15d3b7b77ff3df3e2451ceff00adafad71c26

GPDB merge conflicts:
There was a conflict where the original patch assumed that we had pending
sync request refactor which introduced the function
`register_forget_request`. GPDB 6X_STABLE does the sync request cancel
before the truncate. See this stackflow for more information:
CommitTransaction => smgrDoPendingDeletes => smgrdounlinkall => mdunlink
=> ForgetRelationFsyncRequests

Co-authored-by: Jimmy Yih <jyih@vmware.com>
Co-authored-by: Gaurab Dey <gaurabd@vmware.com>

Commit:
Close AO segment file

Previously we only flushed but didn't close AO segment in
AppendOnlyStorageWrite_FlushAndCloseFile() function. It caused
file descriptors stuck in backends' cache, that could delay kernel
to free disk space after such files would be unlinked.

Co-authored-by: Asim R P <pasim@vmware.com>

Annu149 committed:
[Backport 6X]Adding timeout to stop the segment in the testcase.

Sometimes pg_ctl was failing with default timeout resulting into testcase failure. To make sure that the instance is stopped, added timeout of 120 seconds.

(cherry-picked from commit 57b676af4f25c1443dccc19233b1bb685b7dab68)

Sasasu committed:
enable pg_trgm by default

pg_trgm is a build-in extension of PostgreSQL. this patch enables it in Greenplum.

pg_trgm does not contain new file formats or executor changes, which has good compatibility with
Greenplum.

backport from 7X a41af75bf2266be77a7a703fe77143eeabe98370

Signed-off-by: Sasasu <i@sasa.su>

Sasasu committed:
enable btree_gin by default

btree_gin is a build-in extension of PostgreSQL. this patch enables it in Greenplum.

btree_gin does not contain new file formats or executor changes. which has good compatibility with
Greenplum.

backport from 7X d966becbefc2cc384643b0e50a781ce6c9c8f427

Signed-off-by: Sasasu <i@sasa.su>

DellCurry committed:
Fix gpinitsystem/gpconfig comment other GUCs with same prefix.

Fix #12957, modify regex in gp_bash_functions.sh and gpconfig_helper.py from prefix match to full match after removing spaces.

When we use gpconfig to add a new parameter, it may comment another parameter with the same prefix, because in gpconfig_hepler.py line 96, the function comment_parameter() uses startswith to match the GUC.
The same accident also happens in gpinitsystem with -p option, it will call SED_PG_CONF in gp_bash_functions.sh, which also uses prefix match somewhere.
6X_STABLE and other branches will also benefit from this commit.

Authored-by:  DellCurry

Commit:
Remove unnecessary recheck/scalar from Bitmap Scan

Prior to this commit, ORCA would project recheck/scalar filter columns
in addition to the output columns for
BitmapTableScan/DynamicBitmapTableScan. This would work fine for most of
the cases, but would produce wrong results in case of SharedScan with
BitmapTableScan on AO tables. This commit fixes the issue by removing
the unnecessary recheck/scalar filter, similar to how planner projects
for BitmapHeapScan.

This check was added as part of commit https://github.com/greenplum-db/gpdb/commit/97edf9b1cbd47c33918e5780e04ea241b59a2f46
prior to which for AOCO tables, the tuple slot returned to
BITMAPTABLESCAN by the BitmapIndexScan(under it) had only the output
attribute required, but actually it needed more attributes to evaluate
the recheck condition, which would result in passing uninitialized
field(for the recheck/filter condition attributes) in the TupleTableSlot
to the expression evaluator, and that would lead to crash/incorrect
results. That is why this commit, fixed it by projecting the
recheck/scalar condition.  But, with the refactoring of executor code
for GPDB6 as part of https://github.com/greenplum-db/gpdb/commit/db5163474d2deea03e482ed34c2e68ea96c8b88a,
BITMAPTABLESCAN was replaced by BITMAPHEAPSCAN altogether which no
longer needs projecting the recheck/scalar condition attributes(similar
to planner). So your proposed solution of removing
AddBitmapFilterColumns from the output list, seems the path forward.

This commit fixes the issue: https://github.com/greenplum-db/gpdb/issues/12796

kaknikhil committed:
gprecoverseg: Do not delete the cluster for newhost scenario

In order to test the `-p` functionality in gprecoverseg, we create a cluster
with spare hosts and then make one or more of the cluster hosts as unreachable
and use -p to make the segments on thost hosts failover to the spare hosts.
For this to work, we had to bring the cluster back to it's original state for
each example of the test scenario. This was done to make sure examples were not
dependent on each other and that they could start with a clean cluster. To
achieve this, we would delete the cluster using gpdeletesystem and reinitialize
it using gpinitsystem.
The problem with this approach was that any changes made to the cluster like
setting a guc would get removed because of running gpdeletesystem and
gpinitsystem. A better way is to reconnect the down hosts and then make the
segments on the spare hosts fallback to their original hosts.

jimmyyih committed:
Exclude default gpbackup directory from pg_rewind

When running pg_rewind, things that exist on the target segment but
not on the source segment will be removed. In the case of HA when
recovering the preferred primary segment (mirror is active primary),
pg_rewind would try to delete the gpbackup default directory
"backups". Although this directory default is not recommended, it's a
valid path that needs to be excluded to protect backup data.

Note: This "backups" directory follows the same special GPDB exclusion
logic but with a unique difference. We must prevent the top-level
directory "backups" from being processed in the target files AND
exclude it in the source files exclusion check. This is because the
"backups" directory could exist only on the primary segment, only on
the mirror segment, or exist in both locations (which is different
from "pg_log" directory which always exists on both locations).

Backported from GPDB master branch:
https://github.com/greenplum-db/gpdb/commit/1a1863998eeb4ed097e2566383495909af0527ba

Merge conflict:
The pg_rewind TAP tests are not really used so had to translate the
TAP test change over to the regress-style pg_rewind extrafiles test.

jimmyyih committed:
Exclude default gpbackup directory from pg_basebackup

When running pg_basebackup, the default gpbackup directory "backups"
was being needlessly added to the basebackup. Along with that, the
target location's "backups" directory could be fully replaced or even
deleted. In the case of HA, this could end up with loss of backups
when running full recovery (gprecoverseg -F). Although this directory
default is not recommended, it's a valid path that needs to be
excluded to protect backup data.

Note: This "backups" directory follows the same special GPDB exclusion
logic but with a unique difference. We must prevent the top-level
directory "backups" from being added to the list of
directories/files. If we do not, we'll end up with an empty "backups"
directory at the target location (which can result in complete loss of
backup data). We don't use the hack in
`src/bin/pg_basebackup/pg_basebackup.c` because we'll end up with an
empty "backups" directory if the directory does not exist.

Backported from GPDB master branch:
https://github.com/greenplum-db/gpdb/commit/b56a3175eaa4d5a14c653042196ead0ce8e52305

jimmyyih committed:
Refactor how pg_upgrade handles heterogeneous partition tables

The heterogeneous partition table check validates that the attributes
of a root partition and its child partitions were properly
synchronized.  If it detected an anomaly (e.g. root partition has
dropped column references but a child partition does not), the
partition table would be flagged and the user would be advised to
recreate the table with CTAS. The workaround provided is correct but
can be extremely costly in production clusters where partition tables
can be extremely large.

To lower the cost of fixing the issue, we now do much more complex
queries and logic to determine which specific child partitions are
problematic. We now also clearly define which types of
heterogeneous/homogeneous partition tables are valid for upgrade. We
then advise the user to CTAS only the problematic child partitions and
do an exchange partition (the CTAS staging table should be properly
created with or without dropped column references depending if the
child partition gets detected by the invalid or misaligned dropped
column reference check). This workaround is a bit more complex and
requires more concentrated user involvement but the time savings is
pretty big (e.g. CTAS an entire partition table with 5000 child
partitions vs. CTAS a couple child partitions). Also, the problematic
tables file output for this check has been minimalized to be similar
to the Postgres check output (e.g. no "why" context). The context will
be clarified from gpupgrade side.

Accompanying the pg_upgrade change is a required change to how pg_dump
--binary-upgrade outputs dropped column references. When running
pg_dump --binary-upgrade, dropped columns are always preserved to make
sure the catalog matches with the physical data files. For GPDB
partition tables, this gets very complicated because the root
partition delegates whether or not the entire partition table will
have dropped columns or not. To match the pg_upgrade changes, we
simply need to suppress the column drop from the partition table DDL
that will be run on the target upgrade cluster and assume that all the
child partitions will not have dropped columns (either it was already
like that or the child partition has been fixed with a CTAS and
exchange partition). However, if the root partition and all of its
child partitions have the same dropped column reference, we will
continue to output the dropped column in the schema dump as it is a
homogeneous partition table by definition and valid for upgrade. This
solution works because the root partition does not store any data so
now only the flagged child partitions (via pg_upgrade --check) need to
be fixed instead of touching the entire partition table.

Note: The pg_dump --binary-upgrade issue won't be a problem for GPDB
7+ because child partition DDL follows upstream Postgres logic where
they are created separately and attached to the root/subroot.

adam8157 committed:
Add interconnect test for RHEL 8

Commit:
Docs - reverting coordinator terminology for 6.x only in newly-converted installation guide markdown

mperezfuster committed:
Docs: modified parameter settings for certain GUCs (#12831)

mperezfuster committed:
Docs: specified that skipvalidation option needs to be used when changing guc gp_add_column_inherits_table_setting (#12834)

huansong committed:
pageinspect bitmap functions

Add pageinspect functions and tests for bitmap meta page, LOV page
items and bitmap pages.

Features that can be added later: bm_page_stats(), similar to
bt_page_stats().

Co-authored-by: Soumyadeep Chakraborty <soumyadeep2007@gmail.com>

(cherry picked from commit 6896b2e4bf0e8e4e9b8086820cbb29a9d497d070)

x4m committed:
Enable replay of AO WAL for standby and archive recovery

AO WAL records were replayed only for standby but not for archive
recovery case. So, if restoring GPDB cluster from archive the AO table
data will be lost. This commit fixes that aspect, now only avoids
replay of AO WAL for crash recovery as for that case we perform fsync
of files and hence are crash consistent. AO WAL will be replayed for
archive recovery as well.

Commit message was written by Ashwin Agrawal.

Reviewed-by: Ashwin Agrawal <aashwin@vmware.com>
Reviewed-by: Jimmy Yih <jyih@vmware.com>

interma committed:
Fix compile warnings which imported in "synchronization of amgetbitmap definition with upstream" (#13013)

* fix compile warnings which imported in "synchronization of amgetbitmap
definition with upstream"
* fix compile warning "warning: 'MemoryContextContains' defined but not
used" in mcxt.c

Commit:
docs - add new guc gp_log_resqueue_priority_sleep_time to ref guide (#13023)

* docs - add new guc gp_log_resqueue_priority_sleep_time to ref guide

* clarify, address comments from bhuvnesh

Commit:
docs - adjust the /etc/sysctl.conf kernel.sem setting (#13027)

Commit:
gpcheckcat orphaned_toast_tables fails when there is temp toast table exist

gpcheckcat failed to handle when there is a temp toast table exists
In current gpcheckcat orphaned_toast_tables querey include toast table with format
<pg_toast.pg_toast_>, updating querey to handle tost table with format
<pg_toast<_temp_*>.pg_toast_>.

zhaorui20005 committed:
Fix gpfdist writable external table false success. (#12965)

###Background
When writing an external table of gpfdist protocol, segments will send 
tuples of data to gpfdist through http protocol. For each segment, it
sends a start post request with "X-GP-SEQ" set to 1 as a start request
with no data. Then it will send several post requests with sequence number
of "X-GP-SEQ".
```azure
[segment] ---->> start post request with header "X-GP-SEQ:1"
Then
    [segment] ---->> data post request with header "X-GP-SEQ:2"
    [segment] ---->> data post request with header "X-GP-SEQ:3"
    ...
    [segment] ---->> data post request with header "X-GP-SEQ:n"
```
But in production environment, the socket between the segment and gpfdist 
in one post request will fail because of the network environment. The 
return value of the syscalls of "send" or "recv" through the failing socket
will be -1.
```azure
    [segment] ---->> data post request with header "X-GP-SEQ:x"  failed
```
At the gpfdist side, it will detect the error and try to send an error msg
back to the segment to inform an error. In normal scenario, the error can
be received by segment which acts as a QE and will report an error to QD so
that the psql will report an error happened during the writing procedure.
But in this network failing situation, the socket is corrupted, any syscall 
through this failed socket will return -1. Then gpfdist will close the file
stream. So any following writing post request to this file stream will be 
ended without any error.
```azure
    [segment] ---->> retry post request with header "X-GP-SEQ:x"  failed
    [segment] ---->> retry post request with header "X-GP-SEQ:x"  failed
    [segment] ---->> retry post request with header "X-GP-SEQ:x"  failed
    [segment] ---->> retry post request with header "X-GP-SEQ:x"  success
```
So at the segment side, no http error code is received but only one socket
error is detected. It will retry through libcurl API "curl_easy_perform" to
resend the post request. And libcurl API will create a new socket and 
connect to the gpfdist listen socket and then retry to send the tuples data.
After many times of tries the network is ok again, then the following post 
requests are sent to gpfdist without any errors reported.
```azure
    [segment] ---->> retry post request with header "X-GP-SEQ:x"  failed
    [segment] ---->> retry post request with header "X-GP-SEQ:x"  success
    [segment] ---->> data post request with header "X-GP-SEQ:x+1" success
        ...
    [segment] ---->> data post request with header "X-GP-SEQ: n"  success
```
There are point of this problem:

- HTTP ERROR msg cannot be sent back to segment
- The file stream has been closed, any following post requests will be closed directly

For the first point cannot be controlled because of the network environment.
But for the second point, we should add an error flag after file stream has
been closed for this reason. 

###Solution
So the solution is to add an error flag when gpfdist tries to send http 
response back to the segments in writable situation. When the "send" fails
because of the network failure, it will record the error and then send 
back "FDIST_INTERNAL_ERROR" to any following post request of this session.

###Step to reproduce
Unfortunately it cannot be easily to reproduce this bug in experimental 
environment by using "curl" or "iptables". Any interference with process
of the socket will cause the socket syscall return 0 but not -1. So the
only way to reproduce is to gdb the gpfdist and set the return value of
"recv" to -1 and then continue. So there is no way to add regress cases
for this commit.



Cherry-pick from 0b63c6e6301f75d30f98b071e276e0c5b38bfb71

Annu149 committed:
[Backport 6X] Undo the banner changes on behave tests once the test has completed.

gpinitsystem Testcases:
Scenario: gpinitsystem succeeds if there is banner on host

Scenario: gpinitsystem succeeds if there is multi-line banner on host
gpinitstandby Testcase:
Scenario: gpinitstandby should not throw error when banner exists on the host

gprecoverseg testcase:
Scenario: gprecoverseg should not return error when banner configured on host

* Added scenario tag 'backup_restore_bashrc' which will backup and restore bashrc file to undo the banner changes on host.

* Changes in backup_bashrc and restore_bashrc functions to pickup the home directory from environment and other minor changes.

* Added error string in restore_bashrc() function as well.

dgkimura committed:
Add correlated subquery testcases

(cherry picked from commit e8ebcf6f6643e82004718d5171c61214d097e86b)

dgkimura committed:
Pass subquery context through callstack

ESubqueryCtxt indicates whether a subquery appears in the project list
(EsqctxtValue) or comparison predicate (EsqctxtFilter).  Currently we
can incorrectly recalculate the context inside PexprSubqueryPred().

Following query incorrectly determined the subquery was in the
comparison predicate:
    ```
    SELECT (SELECT 1)=ALL(SELECT generate_series(1,2));
    ```

Using the incorrect subquery context can cause ORCA to create a plan
that incorrect creates an inner (instead of left outer) correlated apply
operator. Ultimately, that mistake can lead to wrong results.

Fix is the pass the subquery context when available rather than
recalculate.

(cherry picked from commit 706e15d819edea3e00771ba0059757ed1dc8cc10)

dgkimura committed:
Restrict predicate push down on left outer correlated apply

Given the following query:
```
SELECT 3 = ALL (SELECT generate_series(2, 3)) FROM (values (1),(2)) v(a);
```

Without predicate push down:
```
Physical plan:
+--CPhysicalComputeScalar
   |--CPhysicalCorrelatedLeftOuterNLJoin
   |  |--CPhysicalConstTableGet Columns: ["column1" (0)] Values: [(1); (2)]
   |  |--CPhysicalComputeScalar
   |  |  |--CPhysicalConstTableGet
   |  |  +--CScalarProjectList
   |  |     |--CScalarProjectElement "ColRef_0004" (4)
   |  |     |  +--CScalarConst (1)
   |  |     +--CScalarProjectElement "generate_series" (2)
   |  |        +--CScalarFunc (generate_series)
   |  |           |--CScalarConst (2)
   |  |           +--CScalarConst (3)
   |  +--CScalarCmp (=)
   |     |--CScalarConst (3)
   |     +--CScalarIdent "generate_series" (2)
   +--CScalarProjectList
      +--CScalarProjectElement "?column?" (3)
         +--CScalarIdent "ColRef_0004" (4)
```

With predicate push down:
```
Physical plan:
+--CPhysicalComputeScalar
   |--CPhysicalCorrelatedLeftOuterNLJoin
   |  |--CPhysicalConstTableGet Columns: ["column1" (0)] Values: [(1); (2)]
   |  |--CPhysicalComputeScalar
   |  |  |--CPhysicalFilter
   |  |  |  |--CPhysicalComputeScalar
   |  |  |  |  |--CPhysicalConstTableGet
   |  |  |  |  +--CScalarProjectList
   |  |  |  |     +--CScalarProjectElement "generate_series" (2)
   |  |  |  |        +--CScalarFunc (generate_series)
   |  |  |  |           |--CScalarConst (2)
   |  |  |  |           +--CScalarConst (3)
   |  |  |  +--CScalarCmp (=)
   |  |  |     |--CScalarConst (3)
   |  |  |     +--CScalarIdent "generate_series" (2)
   |  |  +--CScalarProjectList
   |  |     +--CScalarProjectElement "ColRef_0004" (4)
   |  |        +--CScalarConst (1)
   |  +--CScalarConst (1)
   +--CScalarProjectList
      +--CScalarProjectElement "?column?" (3)
         +--CScalarIdent "ColRef_0004" (4)
```

Both of these plans result in a SUBPLAN node of type ALL_SUBLINK. Inside
ExecScanSubPlan() executor will check that all evaulated expressions are
true. Issue is that if we push the predicate down and filter out all the
expressions that evaluate to false, then in example query we would
incorrectly return true from ExecSubPlan().

Fix is to avoid pushing down the predicate in the case of SUBPLAN of
type ALL_SUBLINK.

(cherry picked from commit 228767d9e9e61426dd3f830e22b252c48ca5486b)

dgkimura committed:
Allow subplan test to contain scalar ident as left expression

In order to support subplan test expressions with scalar ident params on
the left side of the test expression, we must allow more than one column
to be projected from the restricted result node beneath.

(cherry picked from commit 5ee66202abbf13b422c7ae7af29335bd6944b81e)

lenapivotal committed:
Docs - add GRANT in temp tablespaces guc text (#12993)

* edits the vertices/edges graph and makes correction

* makes small change to temp_tablespace guc description

* edits from peer review

Commit:
docs - rename some ANALYZE ref page varnames to clarify (#13037)

* docs - rename some ANALYZE varnames to clarify

* add _name to the new variable refs

Commit:
Docs - dita cleanup in preparation to convert to markdown (#13052)

* Re-work conditions as plain text

* Remove code styling of xrefs

* Remove styling nested within codeblocks

Commit:
Docs - replace DITA source for Best Practices with markdown source (#13054)

* Re-work conditions as plain text

* Remove code styling of xrefs

* Remove styling nested within codeblocks

* Removing best practices DITA source; adding markdown source with book changes

* Change existing DITA xrefs into best practices to html references

Commit:
Docs - update best practices subnav to use 6.19 path

Aegeaner committed:
(6X backport) Make svec_l2_eq strict function to handle NULL properly.

In `gp_sparse_vector` extension, some operator functions are not defined strict
(e.g. `svec_l2_xx` series comparison functions based on L2 Norm), so they
cannot handle NULL value well and may cause a panic. This PR fixed this issue,
and make the ICW pipeline to run regression tests in `gp_sparse_vector` module,
which were not covered by the pipeline check before.

kainwen committed:
Fix ERRORDATA_STACK_SIZE exceeded issue.

Previous commit 87b9c8f9d introduces a bug in the following code
in the function cdbpullup_findEclassInTargetList:

	Oid hash_proc;

	PG_TRY();
	{
		hash_proc = cdb_hashproc_in_opfamily(hashOpFamily, em->em_datatype);
	}
	PG_CATCH();
	{
		hash_proc = InvalidOid;
	}
	PG_END_TRY();

However, cdb_hashproc_in_opfamily might throw error and the
PG_CATCH does not clean up thus might lead to ERRORDATA_STACK_SIZE
exceeded.

This commit fixes the issue by introducing a version of cdb_hashproc_in_opfamily
that returns InvalidOid instead of throwing error when not
found. Since cdb_hashproc_in_opfamily is extern function, directly
change it might lead to ABI broken.  This new API let the caller to
decide what to do and seems useful.

piyushc01 committed:
Changes to avoid gpinitsystem failure when multiple address interfaces and mirrors. (#13049) (#13063)

This regression was introduced while fixing bash banner issue 248c1c740b2f5f351ef6ee6cc4c1317397c857f9
Due to this output of ifaddrs command is interpreted as text instead of array which in turn
resulted in some ip missing in pg_hba.conf. That led to failure of gpinitsystem when multiple
interfaces are present and cluster created with mirrors.

huansong committed:
Revert "pageinspect bitmap functions"

This reverts commit de3e8dfa937f4362071bb08b2cec0b4131b76ee1.

We need to hold this one off for 6.20.0 until we've done another
maintenance release for 6.19.x.

khuddlefish committed:
Revert "(6X backport) Make svec_l2_eq strict function to handle NULL properly."

This reverts commit 2b745d4f3df439c5ffef32af51ab6d2f7a08d237.

This caused icw to fail with asserts so reverting until it can be fixed

khuddlefish committed:
Revert "enable btree_gin by default"

This reverts commit a66eeadb43828e8a3af9ff8f9b4b4302608dfdfe.

This adds a feature. Saving for the next minor release

khuddlefish committed:
Revert "enable pg_trgm by default"

This reverts commit 9453df73f99f754a1d96ff8ec84245d1c3b72b1e.

This adds a feature. Saving for the next minor release

SKeyerror committed:
update the error message when OOM happens (#12971) (#13040)

By default, Greenplum tracks virtual memory internally for resource group using a 
memory auditor referred to as vmtracker, those methods defined in vmem_tracker.h.

If we enabled resource group, memory allocation will go through an additional step:
allocate memory from vmtracker first.

To implement this, Greenplum defines MemoryAllocationStatus to represent the state 
of memory allocation:

- MemoryAllocation_Success
- MemoryFailure_VmemExhausted
- MemoryFailure_SystemMemoryExhausted
- MemoryFailure_QueryMemoryExhausted
- MemoryFailure_ResourceGroupMemoryExhausted

When OOM happens, Greenplum should output different logs according to the state of 
memory allocation. But for now, the error message is confusing and misleading.

So this pr is try to make it clear what things really going on to users when OOM happens.

Co-authored-by: Zhenglog Li <lzhenglong@vmware.com>

BaiShaoqi committed:
Run rhel8 ICW jobs in specific workers

To help with load balancing

[GPR-639]

Authored-by: Shaoqi Bai <bshaoqi@vmware.com>

Commit:
Docs - dita changes to facilitate md conversion for Administration Guide (#13060)

* Docs - dita changes to facilitate md conversion

* More xml cleanup, cleanup around gpcopy to fix oss book build

Commit:
docs - clarify the use and value of diskquota.max_active_tables cfg p… (#13059)

* docs - clarify the use and value of diskquota.max_active_tables cfg param

* tables -> relations with example types

huansong committed:
Add a test for commit "Fix memory leak added by greenplum-db@1805bbf"

The test examines memory usage of MdCxt before and after checkpoint
to see if there's a memory leak.

tvar committed:
Fix memory leak added by https://github.com/greenplum-db/gpdb/commit/1805bbf915df9d7bd07a48d208dda84ddae3626b

huansong committed:
Existing reloption shouldn't affect redistribution in ALTER TABLE SET DISTRIBUTE

This fixes an issue where existing reloptions would cause a reorganize in
situations that it should not. We used to allow ALTER TABLE SET DISTRIBUTE WITH
clause to have additional storage options, and when they are present, we
will re-distribute data over the segments.

However since 50f2e3bbb88, we don't allow new reloptions to be supplied in the
ALTER TABLE SET DISTRIBUTED WITH clause. So this commit is really a follow-up,
removing redistribution decision making logic based on reloptions.
Also renamed new_rel_opts() to get_rel_opts() accordingly.

Cherry-picked-from: 71664c2aa880bef52fdd6db391eb6b034e4a92d2

l-wang committed:
GPORCA CI: add workload6 to explain pipeline

x4m committed:
Fix typo in DeserializeResGroupInfo().

len was used instead of cpuset_len. In most of cases this resulted it
garbage being copied. But rare out-of-page reades could cause a crash.

kainwen committed:
Fix getResUsage integer overflow.

In the function getResUsage it will set the group Id (type Oid)'s
value from a string. Previous code use pg_atoi to do the value parse,
this is not correct because type Oid is uint, we should use
atooid. Another place is building the SQL involving group id it uses
"%d", this commits fix this by using "%u".

(cherry picked from commit 29a9f92fcbb3fd21ddbd7fe0bb61c9abb0d28457)

Commit:
Do not download VC_redist.x64.exe in windows client test (#13079)

VC_redist.x64.exe is downloaded and saved into Windows terraform image
snapshot. Use the local copy to avoid downloading it every time which
may fail sometimes because of network issue.

sambitesh committed:
Introducing parallel array_agg (#13051)

An aggregate to be eligible for multi-stage agg, it needs to have combine
function, serialize and deserialize functions defined. We introduce a parallel
version of `array_agg` called `gp_array_agg`. Since we cannot make catalog
changes in 6X so we are introducing it as an extension. In 7X this will be part
of catalog itself.

chrishajas committed:
Optimize retrieving indexes on partitions within Orca (#13077)

This optimizes the common case for remapping of leaf partitions to root
partitions for use in Orca. In recordIndexesOnLeafPart, we don't need to
recompute the attamp for each index, we just need to do it once. For one
query with many indexes and columns on a partitioned table, this brought
the planning time from 7.5s to .85s using Orca.

In varattnos_map, we would check each attribute of the old and new tuple
descriptor in an n^2 fashion. This is necessary to handle dropped/moved
columns. However, instead we'll do a first iteration through all the
columns, if they're equal and the root/leaf partition attributes are
identical, we don't need to do this additional n^2 logic. This brings
the planning time for this query down from .85s to .4s and this attmap
remapping portion to <50ms.

VMWareJennyBeth committed:
DOCS: Correct locale option default value (#13098)

* Correct locale option default value

* Remove lingering phrase

VMWareJennyBeth committed:
Document that region must match Minio server Region setting (#13097)

deart2k committed:
ADBDEV-2504 ADB 6.19.2 sync

Stolb27 commented:
Interesting commits:

* [b5de722](https://github.com/arenadata/gpdb/pull/306/commits/b5de722ce37c4a7b2b80a784b1de623c092ee2fc) fix for AO tables recovery
* [a39ff3d](https://github.com/arenadata/gpdb/pull/306/commits/a39ff3d53fcd831a6efbc9d2ebf55706ebadbfda) port of commit by @InnerLife0 
* [8c33cb3](https://github.com/arenadata/gpdb/pull/306/commits/8c33cb3c8f41b2dadd11e10762e0440e7a205417) fix for compiller warnings implemented by our patch
* [9190aa5](https://github.com/arenadata/gpdb/pull/306/commits/9190aa54e733c0bf0d7173bf895ac3704be41c10) and [34db8a6](https://github.com/arenadata/gpdb/pull/306/commits/34db8a6927a57aad49f1dd45532ed9846f6e69c4) gpbackup default backup path is excluded from copying during segment recovery. it's genius.
* [af83009](https://github.com/arenadata/gpdb/pull/306/commits/af830092c7567f5baa156019dd2504cc059554a1) important moment about PG_TRY/PG_CATCH
* [5f7e3ed](https://github.com/arenadata/gpdb/pull/306/commits/5f7e3ed9a52637293c95939bcf356e12303a59c0) fix for memory leak on mirrors

InnerLife0 committed:
Fix segfault on execution of multilevel correlated queries. (#12124)

Execution of multilevel correlated queries with a high level of nesting
can cause segfault(when using array_agg, json_agg) or can provide wrong
results (when using classic aggs like sum()). Due to some GP
limitations, correlated subqueries with skip-level correlations are not
supported. Additional check condition is provided to prevent such
queries from planning. QueryHasDistributedRelation function, used by
this check, doesn't recurse over subplans and may return wrong results
for distributed RTE_RELATION entries hidden by RTE_SUBQUERY entries.

Commit fixes such behavior by adding optional recursion to
QueryHasDistributedRelation function. Additional regression test is
included. Additional information can be found at issue #12054.

Stolb27 committed:
fix subselect_gp test on power

apply ebb691a03fb959444d093763e26f5f67a6cc7954 to our custom power answer file

git show ebb691a03fb959444d093763e26f5f67a6cc7954 -- src/test/regress/expected/subselect_gp.out > /tmp/subselect.patch
patch -p1 --merge src/test/regress/expected/subselect_gp_1.out < /tmp/subselect.patch

Stolb27 committed:
Merge branch 'adb-6.x' into 6.19.2-sync

InnerLife0 reviewed (dismissed):
All OK if CI will pass.

Stolb27 commented:
@InnerLife0, @maksm90, problem with power tests described in https://arenadata.atlassian.net/browse/ADBDEV-2539. I think, it's separate problem isn't connected with this sync.

z-wenlin committed:
Fix flaky failed cases pg_rewind_fail_missing_xlog. (#13025)

The case is tesing the bug when gp crash recovery, it calls
two checkpoints, if some checkpoint triggered removed/recylced
wal file, the gp_rewind might failed.

The flaky is due to sometimes when stoping the primary and
promoting the mirror, the end of recovery hasn't fully completed
after mirror is promoted, then insert some data will see errors
"seg is in recovery mode".

The fix is waiting for the end of recovery CHECKPOINT completed
after the mirror was promoted.

khuddlefish committed:
Revert "Introducing parallel array_agg (#13051)"

This reverts commit 2fb13d7facd56ea13d49a8bb330dc8a52c5bfa47.

This feature was merged prematurely. Will bring it back later.

x4m committed:
Install pg_attribute_encoding for AOCS during CTAS (#12937)

When the table is created using "CREATE TABLE ... AS ... WITH NO DATA" 
pg_attribute_encoding may be omitted in some cases. This can result in a 
crash when reading the table with AOCS storage.

This commit fixes installment of pg_attribute_encoding for CTAS WITH NO 
DATA and adds a test for the case.

Fix for #12936.

Co-authored-by: Heikki Linnakangas <hlinnakangas@pivotal.io>
Co-authored-by: Andrey M. Borodin <x4mmm@flight.local>

InnerLife0 committed:
Fix index corruption when invalid snapshot used with AO tables. (#12797)

Using of regular MVCC snapshot to access AO-tables metadata may cause index building (performed globally) inconsistency, because tuples inserted (the same globally) at another opened transactions are not visible. Using the same MVCC snapshot to access AO-tables data ignore RECENTLY_DEAD tuples, used, for example, by "repeatable read" transactions, which again leads to inconsistent results caused by index access.

Commit fixes such behavior using SnapshotSelf for metadata access and SnapshotAny for data access. Additional isolation tests shows how to reproduce the bug before the fix.

huansong committed:
Revert "Fix index corruption when invalid snapshot used with AO tables. (#12797)"

This reverts commit 44dc41a212b98b7b60c470f8f85ffc72894526fe.

This introduced a compile error on pipeline when -Werror=maybe-uninitialized
is given. We somehow missed the pipeline check in the original PR.

The fix should be simple but since this is blocking release 6.19.3 now and
it takes time to verify the fix on pipeline, we'll just revert it and
re-submit a fixed version of the commit.

xiaoxiaoHe-E committed:
set staging table schema to external-schema in yaml file (#13109)

xiaoxiaoHe-E committed:
fix gpload causing pipeline fail (#13131)

Stolb27 committed:
Merge tag '6.19.3' into 6.19.3-sync

6.19.3

Stolb27 committed:
ADBDEV-2504 ADB 6.19.3 sync
